crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -1284,6 +1286,199 @@ impl QueueStorage {
+                } = item_queue_info;
+
+                let thumbnail_info = if let Some((
+                    thumbnail_info,
+                    thumbnail_media_request,
+                    thumbnail_content_type,
+                )) = thumbnail
+                {
+                    let upload_thumbnail_txn = thumbnail_info.txn.clone();
+
+                    // Save the thumbnail upload request as a dependent request of the last file
+                    // upload.
+                    store
+                        .save_dependent_queued_request(
+                            &self.room_id,
+                            &last_upload_file_txn.unwrap(),
+                            upload_thumbnail_txn.clone().into(),
+                            created_at,
+                            DependentQueuedRequestKind::UploadFileOrThumbnail {
+                                content_type: thumbnail_content_type.to_string(),
+                                cache_key: thumbnail_media_request.clone(),
+                                related_to: send_event_txn.clone(),
+                                parent_is_thumbnail_upload: false,
+                            },
+                        )
+                        .await?;
+
+                    // Save the file upload request as a dependent request of the thumbnail upload.
+                    store
+                        .save_dependent_queued_request(
+                            &self.room_id,
+                            &upload_thumbnail_txn,
+                            upload_file_txn.clone().into(),
+                            created_at,
+                            DependentQueuedRequestKind::UploadFileOrThumbnail {
+                                content_type: content_type.to_string(),
+                                cache_key: file_media_request.clone(),
+                                related_to: send_event_txn.clone(),
+                                parent_is_thumbnail_upload: true,
+                            },
+                        )
+                        .await?;
+
+                    Some(thumbnail_info)
+                } else {
+                    // Save the file upload as a dependent request of the last file upload.
+                    store
+                        .save_dependent_queued_request(
+                            &self.room_id,
+                            &last_upload_file_txn.unwrap(),
+                            upload_file_txn.clone().into(),
+                            created_at,
+                            DependentQueuedRequestKind::UploadFileOrThumbnail {
+                                content_type: content_type.to_string(),
+                                cache_key: file_media_request.clone(),
+                                related_to: send_event_txn.clone(),
+                                parent_is_thumbnail_upload: false,
+                            },
+                        )
+                        .await?;
```

Comment:
This is always done in both branches, so it can be extracted out of the if/else and be put just after it (by passing `parent_is_thumbnail_upload`, and mutating `last_upload_file_txn` in place in the thumbnail case). Less code FTW!